File: apps/android/test/domain/sync_manifest_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:castwright/src/domain/sync_manifest.dart';

void main() {
  group('SyncManifestIndex', () {
    test('parses the index shape with books and the full active set', () {
      final index = SyncManifestIndex.fromJson({
        'schemaVersion': 1,
        'books': [
          {
            'bookId': 'b1',
            'updatedAt': '2026-06-06T10:00:00.000Z',
            'title': 'Book One',
            'author': 'Anon',
            'series': 'Saga',
            'seriesPosition': 2,
            'chapterCount': 12,
            'coverUrl': '/api/books/b1/cover',
          },
        ],
        'activeBookIds': ['b1', 'b2'],
      });

      expect(index.schemaVersion, 1);
      expect(index.activeBookIds, ['b1', 'b2']);
      expect(index.books, hasLength(1));
      final b = index.books.single;
      expect(b.bookId, 'b1');
      expect(b.updatedAt, '2026-06-06T10:00:00.000Z');
      expect(b.title, 'Book One');
      expect(b.author, 'Anon');
      expect(b.series, 'Saga');
      expect(b.seriesPosition, 2);
      expect(b.chapterCount, 12);
      expect(b.coverUrl, '/api/books/b1/cover');
    });

    test('tolerates a null seriesPosition and an absent coverUrl', () {
      final index = SyncManifestIndex.fromJson({
        'schemaVersion': 1,
        'books': [
          {
            'bookId': 'b1',
            'updatedAt': '2026-06-06T10:00:00.000Z',
            'title': 'Loose',
            'author': '',
            'series': '',
            'seriesPosition': null,
            'chapterCount': 0,
          },
        ],
        'activeBookIds': ['b1'],
      });

      final b = index.books.single;
      expect(b.seriesPosition, isNull);
      expect(b.coverUrl, isNull);
    });
  });

  group('SyncManifestIndexBook', () {
    test('fromJson parses finished + hidden (default false when absent)', () {
      final b = SyncManifestIndexBook.fromJson({'bookId': 'b1', 'updatedAt': 't', 'finished': true});
      expect(b.finished, isTrue);
      expect(b.hidden, isFalse);
    });
  });

  group('SyncManifestBookDetail', () {
    test('parses uuid-keyed chapters with the full active set', () {
      final detail = SyncManifestBookDetail.fromJson({
        'schemaVersion': 1,
        'bookId': 'b1',
        'updatedAt': '2026-06-06T10:00:00.000Z',
        'chapters': [
          {
            'uuid': 'u-aaa',
            'id': 1,
            'title': 'Chapter 1',
            'fingerprint': '2026-06-06T09:00:00.000Z|123456',
            'urlSuffix': 'audio.m4a',
            'audioUrl': '/api/books/b1/chapters/1/audio.m4a',
            'durationSec': 321.5,
            'lufs': -18.2,
          },
          {
            'uuid': 'u-bbb',
            'id': 2,
            'title': 'Chapter 2 (no audio yet)',
          },
        ],
        'activeChapterUuids': ['u-aaa', 'u-bbb'],
      });

      expect(detail.bookId, 'b1');
      expect(detail.activeChapterUuids, ['u-aaa', 'u-bbb']);
      expect(detail.chapters, hasLength(2));

      final c1 = detail.chapters.first;
      expect(c1.uuid, 'u-aaa');
      expect(c1.id, 1);
      expect(c1.title, 'Chapter 1');
      expect(c1.fingerprint, '2026-06-06T09:00:00.000Z|123456');
      expect(c1.urlSuffix, 'audio.m4a');
      expect(c1.audioUrl, '/api/books/b1/chapters/1/audio.m4a');
      expect(c1.durationSec, 321.5);
      expect(c1.lufs, -18.2);
      expect(c1.hasAudio, isTrue);

      final c2 = detail.chapters.last;
      expect(c2.uuid, 'u-bbb');
      expect(c2.fingerprint, isNull);
      expect(c2.audioUrl, isNull);
      expect(c2.hasAudio, isFalse);
    });

    test('expectedSize is parsed from the fingerprint size component', () {
      final ch = SyncManifestChapter.fromJson({
        'uuid': 'u',
        'id': 3,
        'title': 't',
        'fingerprint': '2026-06-06T09:00:00.000Z|987',
        'urlSuffix': 'audio.mp3',
        'audioUrl': '/api/books/b1/chapters/3/audio.mp3',
      });
      expect(ch.expectedSize, 987);
    });

    test('expectedSize is null when there is no fingerprint', () {
      final ch = SyncManifestChapter.fromJson({
        'uuid': 'u',
        'id': 3,
        'title': 't',
      });
      expect(ch.expectedSize, isNull);
    });
  });
}

File: server/src/workspace/sync-manifest.test.ts
/* srv-32 (plan 191) — pure sync-manifest builders. No I/O. */

import { describe, it, expect } from 'vitest';
import {
  chapterFingerprint,
  bookManifestUpdatedAt,
  buildSyncManifestIndex,
  buildSyncManifestBookDetail,
  type ChapterAudioFact,
} from './sync-manifest.js';
import type { BookStateJson } from './scan.js';

type Chapter = BookStateJson['chapters'][number];

function ch(p: Partial<Chapter> & { id: number; uuid: string }): Chapter {
  return {
    title: `Chapter ${p.id}`,
    slug: `${String(p.id).padStart(2, '0')}-chapter-${p.id}`,
    ...p,
  };
}

function state(over: Partial<BookStateJson> & { chapters: Chapter[] }): BookStateJson {
  return {
    bookId: 'b1',
    manuscriptId: 'm',
    title: 'Book One',
    author: 'Author',
    series: 'Standalones',
    seriesPosition: null,
    isStandalone: true,
    manuscriptFile: 'manuscript.txt',
    castConfirmed: true,
    coverGradient: ['#000', '#fff'],
    createdAt: '2026-01-01T00:00:00.000Z',
    updatedAt: '2026-01-01T00:00:00.000Z',
    ...over,
  };
}

describe('chapterFingerprint', () => {
  it('is undefined without an audioRenderedAt or a file size', () => {
    expect(chapterFingerprint(undefined, 100)).toBeUndefined();
    expect(chapterFingerprint('2026-01-01T00:00:00.000Z', undefined)).toBeUndefined();
  });

  it('combines audioRenderedAt and file size into a string', () => {
    const fp = chapterFingerprint('2026-01-01T00:00:00.000Z', 4096);
    expect(typeof fp).toBe('string');
    expect(fp).toContain('4096');
  });

  it('changes when audioRenderedAt changes (every audio mutation bumps it)', () => {
    const a = chapterFingerprint('2026-01-01T00:00:00.000Z', 4096);
    const b = chapterFingerprint('2026-02-02T00:00:00.000Z', 4096);
    expect(a).not.toBe(b);
  });

  it('changes when the file size changes', () => {
    const a = chapterFingerprint('2026-01-01T00:00:00.000Z', 4096);
    const b = chapterFingerprint('2026-01-01T00:00:00.000Z', 8192);
    expect(a).not.toBe(b);
  });
});

describe('bookManifestUpdatedAt', () => {
  it('falls back to state.updatedAt when no chapter has audio', () => {
    const s = state({
      updatedAt: '2026-03-01T00:00:00.000Z',
      chapters: [ch({ id: 1, uuid: 'u1' })],
    });
    expect(bookManifestUpdatedAt(s)).toBe('2026-03-01T00:00:00.000Z');
  });

  it('returns the latest audioRenderedAt when it is newer than state.updatedAt', () => {
    const s = state({
      updatedAt: '2026-03-01T00:00:00.000Z',
      chapters: [
        ch({ id: 1, uuid: 'u1', audioRenderedAt: '2026-04-10T00:00:00.000Z' }),
        ch({ id: 2, uuid: 'u2', audioRenderedAt: '2026-05-20T00:00:00.000Z' }),
      ],
    });
    expect(bookManifestUpdatedAt(s)).toBe('2026-05-20T00:00:00.000Z');
  });
});

describe('buildSyncManifestIndex', () => {
  const books = [
    {
      bookId: 'b1',
      state: state({ bookId: 'b1', updatedAt: '2026-01-01T00:00:00.000Z', chapters: [ch({ id: 1, uuid: 'u1' })] }),
      coverUrl: '/api/books/b1/cover',
    },
    {
      bookId: 'b2',
      state: state({ bookId: 'b2', updatedAt: '2026-06-01T00:00:00.000Z', chapters: [ch({ id: 1, uuid: 'u2' })] }),
    },
  ];

  it('lists every book with an audio-aware updatedAt and the full activeBookIds set', () => {
    const idx = buildSyncManifestIndex(books);
    expect(idx.books.map((b) => b.bookId).sort()).toEqual(['b1', 'b2']);
    expect(idx.activeBookIds.sort()).toEqual(['b1', 'b2']);
    expect(idx.books.find((b) => b.bookId === 'b1')?.coverUrl).toBe('/api/books/b1/cover');
  });

  it('?since trims the books list but keeps the FULL activeBookIds set', () => {
    const idx = buildSyncManifestIndex(books, '2026-03-01T00:00:00.000Z');
    // only b2 changed after the cutoff
    expect(idx.books.map((b) => b.bookId)).toEqual(['b2']);
    // but the active set still lists every book so the client can evict
    expect(idx.activeBookIds.sort()).toEqual(['b1', 'b2']);
  });

  it('carries finished + hidden flags through the index', () => {
    const idx = buildSyncManifestIndex([
      { bookId: 'b1', state: state({ bookId: 'b1', updatedAt: '2026-01-01T00:00:00.000Z', chapters: [ch({ id: 1, uuid: 'u1' })] }), finished: true, hidden: false },
      { bookId: 'b2', state: state({ bookId: 'b2', updatedAt: '2026-06-01T00:00:00.000Z', chapters: [ch({ id: 1, uuid: 'u2' })] }), hidden: true },
    ]);
    const b1 = idx.books.find((b) => b.bookId === 'b1')!;
    const b2 = idx.books.find((b) => b.bookId === 'b2')!;
    expect(b1.finished).toBe(true);
    expect(b2.hidden).toBe(true);
  });
});

describe('buildSyncManifestBookDetail', () => {
  const s = state({
    bookId: 'b1',
    chapters: [
      ch({
        id: 1,
        uuid: 'u1',
        audioRenderedAt: '2026-04-01T00:00:00.000Z',
        audioQa: {
          status: 'ok',
          reasons: [],
          measuredLufs: -16.2,
          truePeakDb: -1.5,
          durationSec: 123.4,
          expectedSec: 120,
          checkedAt: '2026-04-01T00:00:00.000Z',
        },
      }),
      ch({ id: 2, uuid: 'u2' }), // no audio yet
      ch({ id: 3, uuid: 'u3', excluded: true }), // front/back-matter
    ],
  });
  const audio = new Map<number, ChapterAudioFact>([
    [1, { fileSize: 4096, urlSuffix: 'audio.mp3' }],
  ]);

  it('emits uuid-keyed active chapters with fingerprint/urlSuffix/audioUrl/duration/lufs', () => {
    const detail = buildSyncManifestBookDetail('b1', s, audio);
    expect(detail.bookId).toBe('b1');
    const c1 = detail.chapters.find((c) => c.uuid === 'u1')!;
    expect(c1.id).toBe(1);
    expect(c1.fingerprint).toContain('4096');
    expect(c1.urlSuffix).toBe('audio.mp3');
    expect(c1.audioUrl).toBe('/api/books/b1/chapters/1/audio.mp3');
    expect(c1.durationSec).toBe(123.4);
    expect(c1.lufs).toBe(-16.2);
  });

  it('uses the audio fact durationSec (from the segments file) when audioQa has none', () => {
    const s2 = state({
      bookId: 'b1',
      chapters: [ch({ id: 1, uuid: 'u1', audioRenderedAt: '2026-04-01T00:00:00.000Z' })],
    });
    const audio2 = new Map<number, ChapterAudioFact>([
      [1, { fileSize: 4096, urlSuffix: 'audio.mp3', durationSec: 99.5 }],
    ]);
    const detail = buildSyncManifestBookDetail('b1', s2, audio2);
    expect(detail.chapters[0].durationSec).toBe(99.5);
  });

  it('lists a chapter without audio but with no fingerprint/urlSuffix', () => {
    const detail = buildSyncManifestBookDetail('b1', s, audio);
    const c2 = detail.chapters.find((c) => c.uuid === 'u2')!;
    expect(c2).toBeDefined();
    expect(c2.fingerprint).toBeUndefined();
    expect(c2.urlSuffix).toBeUndefined();
    expect(c2.audioUrl).toBeUndefined();
  });

  it('excludes excluded chapters from both chapters and activeChapterUuids', () => {
    const detail = buildSyncManifestBookDetail('b1', s, audio);
    expect(detail.chapters.map((c) => c.uuid).sort()).toEqual(['u1', 'u2']);
    expect(detail.activeChapterUuids.sort()).toEqual(['u1', 'u2']);
  });
});
